DDL design document (#17266)
hazelcast/hazelcast: docs/design/sql/05-ddl.md

| Table type | Description |
|---|---|
| IMap | Table backed by Hazelcast `IMap` |
| ReplicatedMap | Table backed by Hazelcast `ReplicatedMap` |
| File | Table backed by files in a directory |

**table_parameter**

Parameter specific to the external storage system - connection string, serialization format etc.
This is an identifier, not a string. Therefore it must not be quoted as a string. It can be quoted in double
quotes if it contains special characters (such as space, period...).

**table_parameter_value**

Parameter value as a string literal.

### 1.4 Examples

- Create an external table referencing an `IMap` with inferred schema.

  ```sql
  CREATE EXTERNAL TABLE persons
  TYPE "com.hazelcast.IMap"
  OPTIONS (
    "serialization.key.format" 'java',
    "serialization.key.java.class" 'java.lang.Integer',
    "serialization.value.format" 'java',
    "serialization.value.java.class" 'java.lang.String'
  )
  ```

  Column names and types are inferred as:

  ```sql
  (
    __key INT,
    this VARCHAR
  )
  ```

- Create an external table referencing an `IMap` with explicit schema.

  ```sql
  CREATE EXTERNAL TABLE persons (
    id INT EXTERNAL NAME __key,
    name VARCHAR
  )

>   TYPE "com.hazelcast.IMap"

Reviewer: Have we agreed to use short names for connectors? If yes, then this part of the document should be adjusted (probably some other places as well)

# SQL DDL

## Overview

This part contains reference information for the DDL commands supported by Hazelcast Mustang.

## Table of Contents
1. [CREATE EXTERNAL TABLE](#1-create-external-table) -- define a new external table
2. [CREATE EXTERNAL TABLE AS](#2-create-external-table-as) -- define a new external table from the results of a query
3. [DROP EXTERNAL TABLE](#3-drop-external-table) -- remove external table

## 1 CREATE EXTERNAL TABLE

`CREATE EXTERNAL TABLE` -- define a new external table

### 1.1 Synopsis

```sql
CREATE [OR REPLACE] EXTERNAL TABLE [IF NOT EXISTS] table_name [ (
 column_name column_type [ EXTERNAL NAME column_external_name ] [, ... ]
) ]
TYPE table_type
[ OPTIONS ( table_parameter table_parameter_value [, ... ] ) ]
```

### 1.2 Description

> `CREATE EXTERNAL TABLE` registers a virtual table that references an external storage system. For some storage systems,
> it does not create a physical entity until a write occurs. Column list is optional, as for some storage systems and
> serialization formats, it is possible to infer it.

Reviewer 2: ```suggestion
`CREATE EXTERNAL TABLE` registers a virtual table that references an external storage system. For some storage systems
it does not create a physical entity until a write occurs. Column list is optional; if not specified, we will attempt
to infer it from sample data or from the options.
```

The table is created in the `public` schema. The name of the table must be distinct from the name of any other table in
the schema.

### 1.3 Parameters

**OR REPLACE**

Replace table definition if one with the same name already exists.

**IF NOT EXISTS**

> Do not fail if a table with the same name already exists.

Reviewer 2: This will be more clear:
```suggestion
Take no action if a table with the same name already exists.
```

**table_name**

> The unqualified name of the table to be created.

Author: I'm not sure, we don't support creating schemas (yet?), so prefixing with `public` would be the only option.

> The unqualified name of the table to be created.

Reviewer 2: Yeah, but schema is allowed in our queries (you can do `SELECT * FROM public.my_map`), so why not allow it here?

**column_name**

The name of the column in the table.

**column_type**

The data type of the column. For more information on the supported data types, refer to [type system document](https://github.com/hazelcast/hazelcast/blob/master/docs/design/sql/01-type-system.md).

**column_external_name**

> The path to the field in the external storage system. Used to link a column to physical field under a different name
> (i.e. to be able to access both key and value `IMap` fields having same name).

Reviewer 2: ```suggestion
The path to the field in the external storage system. Used to link the column to a physical field with a different name.
It's optional, the default value is specific for each table type. For example for IMap it defaults to `this.<columnName>` 
where `this` refers to the entry value.
```

**table_type**

The external storage system identifier. Supported table types are listed in Table 1. Not all storage systems implement
full set of operations, so for instance you might be able to `SELECT` from a given table but not `INSERT` into it.

*Table 1: CREATE EXTERNAL TABLE supported table types*

| Table type | Description |
|---|---|
| com.hazelcast.IMap | Table backed by Hazelcast `IMap` |
| com.hazelcast.ReplicatedMap | Table backed by Hazelcast `ReplicatedMap` |
| com.hazelcast.File | Table backed by files in a directory |

**table_parameter**

> Parameter specific to the external storage system - connection string, serialization format etc.

Reviewer 2: ```suggestion
Parameter specific to the external storage system - connection string, serialization format etc.
This is an identifier, not a string. Therefore it must not be quoted as a string. It can be quoted in double
quotes if it contains special characters (such as space, period...).
```

  ```sql
  CREATE EXTERNAL TABLE persons (
    id INT EXTERNAL NAME "__key.id",
    name VARCHAR EXTERNAL NAME "this.name"
  )
  TYPE "com.hazelcast.IMap"
  OPTIONS (
    "serialization.key.format" 'json',
    "serialization.value.format" 'json'
  )
  ```

## 2 CREATE EXTERNAL TABLE AS

`CREATE EXTERNAL TABLE AS` -- define a new external table from the results of a query

### 2.1 Synopsis

```
CREATE [OR REPLACE] EXTERNAL TABLE [IF NOT EXISTS] table_name [ (
 column_name [ EXTERNAL NAME column_external_name ] [, ... ]
) ]
TYPE table_type
[ OPTIONS ( table_parameter table_parameter_value [, ... ] ) ]
AS query
```

### 2.2 Description

`CREATE EXTERNAL TABLE AS` registers a virtual table that references an external storage system. It fills the table with
data computed by `SELECT` statement. The table columns have the names and data types associated with the output columns
of the `SELECT` (except that column names can be overridden by an explicit list of names).

The table is created in the `public` schema. The name of the table must be distinct from the name of any other table in
the schema.

### 2.3 Parameters

**OR REPLACE**

> Replace table definition if one with the same name already exists.

Author: `DROP` does not clear nor delete the underlying storage system, why `OR REPLACE` should do that?

> Replace table definition if one with the same name already exists.

Reviewer 2: Because it's going to load data into that table. I think `CREATE OR REPLACE` should be equivalent of `DROP` + `CREATE`, except that it's atomic. Also `CREATE TABLE .. AS SELECT` should be equivalent of `CREATE TABLE` + `INSERT ... SELECT`, but atomic. And `CREATE OR REPLACE .. AS SELECT` is `DROP` + `CREATE` + `INSERT .. SELECT`, but atomic. 

In our case the `CREATE + INSERT` part is not atomic, we should make this very clear. But I still think it might be confusing to the users to find out that the old data weren't removed. But thinking about it, even if you don't use the `OR REPLACE` part, the target object can still contain data.

So we should make it more clear that the `CREATE [OR REPLACE]` part only applies to the link and pre-existing data are not cleared. And also that it's not atomic.

### 1.1 Synopsis

```sql
CREATE [OR REPLACE] EXTERNAL TABLE [IF NOT EXISTS] table_name [ (
 column_name column_type [ EXTERNAL NAME column_external_name ] [, ... ]
) ]
TYPE table_type
[ OPTIONS ( table_parameter table_parameter_value [, ... ] ) ]
```

### 1.2 Description

`CREATE EXTERNAL TABLE` registers a virtual table that references an external storage system. For some storage systems,
it does not create a physical entity until a write occurs. Column list is optional, as for some storage systems and
serialization formats, it is possible to infer it.

The table is created in the `public` schema. The name of the table must be distinct from the name of any other table in
the schema.

### 1.3 Parameters

**OR REPLACE**

Replace table definition if one with the same name already exists.

**IF NOT EXISTS**

Do not fail if a table with the same name already exists.

**table_name**

The unqualified name of the table to be created.

**column_name**

The name of the column in the table.

**column_type**

The data type of the column. For more information on the supported data types, refer to [type system
document](01-type-system.md).

> **column_external_name**

Reviewer: Could you please share an example of such an IMap conflict, where both key and value have columns with the same name? How the DDL would look like?

Author: So for instance: 
```sql
  CREATE EXTERNAL TABLE persons (
    key_id INT EXTERNAL NAME "__key.id",
    value_id INT EXTERNAL NAME "this.id"
  )
  TYPE "com.hazelcast.IMap"
  OPTIONS (
    "serialization.key.format" 'java',
    "serialization.key.java.class" 'com.hazelcast.Person',
    "serialization.value.format" 'java',
    "serialization.value.java.class" 'com.hazelcast.Person'
  )
  ```

The path to the field in the external storage system. Used to link a column to physical field under a different name
(i.e. to be able to access both key and value `IMap` fields having same name).

**table_type**

The external storage system identifier. Supported table types are listed in Table 1. Not all storage systems implement
full set of operations, so for instance you might be able to `SELECT` from a given table but not `INSERT` into it.

*Table 1: CREATE EXTERNAL TABLE supported table types*

| Table type | Description |
|---|---|

> | com.hazelcast.IMap | Table backed by Hazelcast `IMap` |

Reviewer: I am not sure we need fully-qualified names here. Why not just "IMAP", "REPLICATED_MAP" and "FILE" instead?
The probability of naming conflict in the connector is very small.

| com.hazelcast.ReplicatedMap | Table backed by Hazelcast `ReplicatedMap` |
| com.hazelcast.File | Table backed by files in a directory |

**table_parameter**

Parameter specific to the external storage system - connection string, serialization format etc.

**table_parameter_value**

Parameter value as a string literal.

### 1.4 Examples

- Create an external table referencing an `IMap` with inferred schema.

  ```sql
  CREATE EXTERNAL TABLE persons
  TYPE "com.hazelcast.IMap"
  OPTIONS (
    "serialization.key.format" 'java',
    "serialization.key.java.class" 'java.lang.Integer',
    "serialization.value.format" 'java',
    "serialization.value.java.class" 'java.lang.String'
  )
  ```

  Column names and types are inferred as:

  ```sql
  (
    __key INT,
    this VARCHAR
  )
  ```

- Create an external table referencing an `IMap` with explicit schema.

  ```sql
  CREATE EXTERNAL TABLE persons (

>     id INT EXTERNAL NAME __key,

Reviewer: Does any other connector except for IMap/ReplicatedMap requires `EXTERNAL NAME`?

Author: Kafka might use it in similar way (to disambiguate key and value fields). Moreover it might be simply used to operate with different names on SQL level (not sure how useful that might be though).

Reviewer 2: For maps and kafka it's required only for key fields. If not present, we assume a value field. But this is type-specific, for example, external JDBC table will never require it, but could be used to rename a column.

Reviewer: Got it. Another question - why do we call it `EXTERNAL NAME`? In terms of Hazelcast, this is a "path", not something external. Basically, we create *aliases* for paths.

Author: External in a sense that it refers to something outside of SQL - i.e. a field in java object in an IMap. Sometimes that would be simply an alias (if you want to just rename) but sometimes a path (if you want to disambiguate). I'm not happy with `external name` but could not find a better name...

Reviewer: It is not a blocker definitely for the whole activity, just sounds a bit strange. Do any other NoSQL products with SQL support have something like this? Maybe we can borrow ideas from them.

A couple of ideas from my side:
1. Maybe we can define a custom function to extract the specific field from the key/value. E.g. `SELECT FIELD('name', __key), name FROM Person` - this first field is taken from the key, the second one is taken from the value
1. Provided that the conflicts are expected to be rare, we may use a custom property instead with a column name attached. E.g. `CREATE TABLE ... OPTIONS path.name=__key.name`. The advantage of this is that the basic syntax is not concerned with connector-specific stuff
